extensions/ql-vscode/src/data-extensions-editor/model-details/model-details-data-provider.ts

```diff
@@ -61,12 +61,45 @@ export class ModelDetailsDataProvider
       return [];
     }
   }
+
+  getParent(
+    item: ModelDetailsTreeViewItem,
+  ): ModelDetailsTreeViewItem | undefined {
+    if (isExternalApiUsage(item)) {
+      return undefined;
+    } else {
+      return this.externalApiUsages.find((e) => e.usages.includes(item));
+    }
+  }
+
+  public resolveCanonicalUsage(usage: Usage): Usage | undefined {
+    for (const externalApiUsage of this.externalApiUsages) {
+      for (const u of externalApiUsage.usages) {
+        if (usagesAreEqual(u, usage)) {
+          return u;
+        }
+      }
+    }
+    return undefined;
+  }
 }
 
-type ModelDetailsTreeViewItem = ExternalApiUsage | Usage;
+export type ModelDetailsTreeViewItem = ExternalApiUsage | Usage;
 
 function isExternalApiUsage(
   item: ModelDetailsTreeViewItem,
 ): item is ExternalApiUsage {
   return (item as any).usages !== undefined;
 }
+
+function usagesAreEqual(u1: Usage, u2: Usage): boolean {
```

Comment:
We could make this a bit cleaner by giving the `Usage` object some unique "key" field similar to `ExternalApiUsage.signature`. The combination of URI and line/column should be unique, so we could use that.

This has the tradeoff that it makes the `Usage` object larger, so we're just passing around more data all of the time, and this is the only operation that it makes any faster.

Reply:
Agree with the approach you took here - let's see how this performs in practise before adding more things and increasing complexity.